src/devices/machine/sis630_gui.cpp

```diff
@@ -0,0 +1,701 @@
+
+	// "SiS 630 (Ver. 2.02.1c) [AGP VGA] (Silicon Integrated Systems Corp.).bin"
+	ROM_SYSTEM_BIOS( 1, "2.02.1c", "Ver. 2.02.1c" )
+	ROMX_LOAD( "sis630.bin", 0x0000, 0x8000, BAD_DUMP CRC(f04ef9b0) SHA1(2396a79cd4045362bfc511090b146daa85902b4d), ROM_BIOS(1) )
+ROM_END
+
+const tiny_rom_entry *sis630_gui_device::device_rom_region() const
+{
+	return ROM_NAME(sis630gui);
+}
+
+void sis630_gui_device::device_add_mconfig(machine_config &config)
+{
+	screen_device &screen(SCREEN(config, "screen", SCREEN_TYPE_RASTER));
+	screen.set_raw(XTAL(25'174'800), 900, 0, 640, 526, 0, 480);
+	screen.set_screen_update(m_svga, FUNC(sis630_svga_device::screen_update));
+
+	SIS630_SVGA(config, m_svga, 0);
+	m_svga->set_screen("screen");
+	// 64MB according to POST
+	// documentation claims 128MB, assume being wrong
+	m_svga->set_vram_size(64*1024*1024);
+}
+
+void sis630_gui_device::config_map(address_map &map)
+{
+	pci_device::config_map(map);
+	map(0x2c, 0x2d).r(FUNC(sis630_gui_device::subvendor_r));
+	map(0x2e, 0x2f).r(FUNC(sis630_gui_device::subsystem_r));
+	map(0x2c, 0x2f).w(FUNC(sis630_gui_device::subvendor_w));
+
+//  map(0x3c, 0x3d) irq line/pin
+
+	map(0x34, 0x34).r(FUNC(sis630_gui_device::capptr_r));
+
+	map(0x50, 0x53).r(FUNC(sis630_gui_device::agp_id_r));
+	map(0x54, 0x57).r(FUNC(sis630_gui_device::agp_status_r));
+	map(0x58, 0x5b).rw(FUNC(sis630_gui_device::agp_command_r), FUNC(sis630_gui_device::agp_command_w));
+	map(0x5c, 0x5c).lr8(NAME([] () { return 0; })); // NULL terminator
+}
+
+u8 sis630_gui_device::capptr_r()
+{
+	return 0x50;
+}
+
+// TODO: move to specific interface
+u32 sis630_gui_device::agp_id_r()
+{
+	LOGAGP("Read AGP ID [$50]\n");
+	// bits 23-16 AGP v1.0
+	// bits 15-8 0x5c NEXT_PTR (which goes to NULL terminator, heh)
+	// bits 7-0 CAP_ID (0x02 for AGP)
+	return 0x00105c02;
+}
+
+u32 sis630_gui_device::agp_status_r()
+{
+	LOGAGP("Read AGP status [$54]\n");
+	// RQ (1 + 1), 2X and 1X capable
+	return 0x01000003;
+}
+
+u32 sis630_gui_device::agp_command_r(offs_t offset, uint32_t mem_mask)
+{
+	LOGAGP("Read AGP command [$58] %d %d %08x\n", m_agp.enable, m_agp.data_rate, mem_mask);
+	// TODO: enable gets cleared by AGP_RESET, or even from PCI RST#
+	return m_agp.enable << 8 | (m_agp.data_rate & 7);
+}
+
+void sis630_gui_device::agp_command_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+{
+	LOGAGP("Write AGP command [$c8] %08x & %08x\n", data, mem_mask);
+
+	if (ACCESSING_BITS_8_15)
+	{
+		m_agp.enable = bool(BIT(m_agp.enable, 8));
+		LOGAGP("- AGP_ENABLE = %d\n", m_agp.enable);
+	}
+
+	if (ACCESSING_BITS_0_7)
+	{
+		std::map<u8, std::string> agp_transfer_rates = {
+			{ 1, "1X" },
+			{ 2, "2X" },
+		};
+
+		// make sure the AGP DATA_RATE specs are honored
+		try {
+			const u8 data_rate = data & 3;
+			LOGAGP("- DATA_RATE = %s\n", agp_transfer_rates.at(data_rate));
+			m_agp.data_rate = data_rate;
+		}
+		catch (std::out_of_range& err) {
+			LOG("Warning: AGP illegal DATA_RATE set = %d enabled=%d\n", data & 3, m_agp.enable);
+		}
```

Comment:
This is a really slow way of doing this – constructing a map causes multiple allocations, and exceptions are slow to throw catch due to needing to page in unwind tables.  It’s not a good look to do things this way in a handler.

Reply:
It’s throwaway code that should be eventually removed and inherited in a common AGP interface, not sure about how to make it pretty & if it’s worth the hassle (it’s a very special case that should probably be an actual emu_exception thrown)